File: third-party/archive/build-packages/build-packages/pf-runner-1.0.1/src/pf-runner/pf_config.py
#!/usr/bin/env python3
"""
pf_config.py - Central config loader for pf-runner and repo scripts.

The project uses a single JSON5 config file (default: pf.config.json5) and avoids
PF_* environment variables for configuration.
"""

from __future__ import annotations

from copy import deepcopy
from pathlib import Path
from typing import Any, Dict, Optional, Tuple

try:
    import json5  # type: ignore
except Exception:  # pragma: no cover - best-effort import for minimal environments
    json5 = None


DEFAULT_CONFIG_FILENAME = "pf.config.json5"

DEFAULT_CONFIG: Dict[str, Any] = {
    "pfy": {
        "file": "Pfyfile.pf",
        "searchParents": "git",  # git|all|none
    },
    "runner": {
        "autocorrect": {
            "mode": "auto",  # auto|ask|off
            "threshold": 0.75,
        },
        "pathAutofix": True,
        "playwright": {
            "headful": False,
        },
    },
    "api": {
        "host": "127.0.0.1",
        "port": 8000,
        "workers": 4,
    },
    "container": {
        "runtime": "podman",
        "image": "localhost/pf-runner:latest",
    },
    "devEnvironment": {
        "useQuadlet": True,
        "gpuSupport": False,
    },
    "os": {
        "distroArtifactsDir": "~/.pf/distros",
        "switchBaseDir": "~/.pf/os-switch",
    },
}


def _deep_merge(base: Dict[str, Any], override: Dict[str, Any]) -> Dict[str, Any]:
    out = dict(base)
    for key, value in override.items():
        if isinstance(value, dict) and isinstance(out.get(key), dict):
            out[key] = _deep_merge(out[key], value)  # type: ignore[arg-type]
        else:
            out[key] = value
    return out


def _git_root(start_dir: Path) -> Optional[Path]:
    cur = start_dir.resolve()
    while True:
        if (cur / ".git").exists():
            return cur
        parent = cur.parent
        if parent == cur:
            return None
        cur = parent


def find_config_path(start_dir: Optional[str] = None, explicit_path: Optional[str] = None) -> Optional[Path]:
    start = Path(start_dir).resolve() if start_dir else Path.cwd().resolve()

    if explicit_path:
        explicit = Path(explicit_path).expanduser()
        if not explicit.is_absolute():
            explicit = (start / explicit).resolve()
        return explicit

    stop = _git_root(start) or Path(start.anchor)
    cur = start
    while True:
        candidate = cur / DEFAULT_CONFIG_FILENAME
        if candidate.is_file():
            return candidate
        if cur == stop or cur.parent == cur:
            break
        cur = cur.parent

    # User config fallbacks
    home = Path.home()
    for candidate in (
        home / ".config" / "pf" / DEFAULT_CONFIG_FILENAME,
        home / ".pf" / DEFAULT_CONFIG_FILENAME,
    ):
        if candidate.is_file():
            return candidate

    # Installed fallback: config alongside pf-runner sources (e.g. /usr/local/lib/pf-runner)
    installed_candidate = Path(__file__).resolve().parent / DEFAULT_CONFIG_FILENAME
    if installed_candidate.is_file():
        return installed_candidate

    return None


def load_config(
    *,
    start_dir: Optional[str] = None,
    explicit_path: Optional[str] = None,
    require_exists: bool = False,
) -> Tuple[Dict[str, Any], Optional[Path]]:
    """
    Load config and return (config_dict, resolved_path).

    If require_exists=True and the resolved path does not exist, raises FileNotFoundError.
    """
    resolved = find_config_path(start_dir=start_dir, explicit_path=explicit_path)
    if resolved is None:
        return deepcopy(DEFAULT_CONFIG), None

    if not resolved.is_file():
        if require_exists:
            raise FileNotFoundError(f"Config file not found: {resolved}")
        return deepcopy(DEFAULT_CONFIG), resolved

    raw = resolved.read_text(encoding="utf-8")
    if json5 is None:
        raise RuntimeError(
            "JSON5 parsing is unavailable (missing python 'json5' package). "
            "Install it or use a strict JSON file."
        )
    parsed = json5.loads(raw)
    if not isinstance(parsed, dict):
        raise ValueError(f"Config root must be an object: {resolved}")

    merged = _deep_merge(deepcopy(DEFAULT_CONFIG), parsed)
    return merged, resolved


def get(cfg: Dict[str, Any], dotted_key: str, default: Any = None) -> Any:
    cur: Any = cfg
    for part in dotted_key.split("."):
        if not isinstance(cur, dict) or part not in cur:
            return default
        cur = cur[part]
    return cur


def get_bool(cfg: Dict[str, Any], dotted_key: str, default: bool) -> bool:
    value = get(cfg, dotted_key, default)
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)):
        return bool(value)
    if isinstance(value, str):
        return value.strip().lower() in {"1", "true", "yes", "on"}
    return default


def get_int(cfg: Dict[str, Any], dotted_key: str, default: int) -> int:
    value = get(cfg, dotted_key, default)
    if isinstance(value, int):
        return value
    if isinstance(value, str):
        try:
            return int(value.strip(), 10)
        except Exception:
            return default
    return default


def get_float(cfg: Dict[str, Any], dotted_key: str, default: float) -> float:
    value = get(cfg, dotted_key, default)
    if isinstance(value, (float, int)):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value.strip())
        except Exception:
            return default
    return default


def resolve_path(value: str, *, base_dir: Optional[Path] = None) -> str:
    expanded = Path(value).expanduser()
    if expanded.is_absolute():
        return str(expanded)
    if base_dir:
        return str((base_dir / expanded).resolve())
    return str(expanded)

File: tools/pf-config.mjs
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

let JSON5;
try {
  // json5 is commonly available via npm or system packages (e.g. /usr/share/nodejs/json5).
  // We keep this as a runtime dependency so repo scripts work without env-var config.
  JSON5 = (await import('json5')).default;
} catch (err) {
  throw new Error(
    "Missing dependency 'json5'. Install it with: npm install json5\n" +
      `Original error: ${err?.message || err}`
  );
}

const DEFAULT_CONFIG = {
  pfy: { file: 'Pfyfile.pf', searchParents: 'git' },
  runner: {
    autocorrect: { mode: 'auto', threshold: 0.75 },
    pathAutofix: true,
    playwright: { headful: false },
  },
  api: { host: '127.0.0.1', port: 8000, workers: 4 },
  container: { runtime: 'podman', image: 'localhost/pf-runner:latest' },
  pr: {
    storageDir: '~/.config/pf',
    discoveryConfigFile: '~/.config/pf/pr-config.json',
    discoveredPrsFile: '~/.config/pf/discovered-prs.json',
    aiProvidersFile: '~/.config/pf/ai-providers.json',
    repositories: [],
    platforms: { github: { enabled: true }, gitlab: { enabled: true } },
    filters: { states: ['open'], labels: [], authors: [] },
    ai: {
      providers: {
        openai: { apiKey: null, model: 'gpt-4', enabled: false },
        anthropic: { apiKey: null, model: 'claude-3-sonnet-20240229', enabled: false },
      },
      reviewCriteria: {
        security: true,
        performance: true,
        maintainability: true,
        testCoverage: true,
        documentation: true,
      },
    },
  },
  github: { repoOwner: 'P4X-ng', repoName: 'pf-web-poly-compile-helper-runner', token: null },
  devEnvironment: { useQuadlet: true, gpuSupport: false },
  os: { distroArtifactsDir: '~/.pf/distros', switchBaseDir: '~/.pf/os-switch' },
};

function deepMerge(base, override) {
  const out = { ...base };
  for (const [key, value] of Object.entries(override || {})) {
    if (
      value &&
      typeof value === 'object' &&
      !Array.isArray(value) &&
      out[key] &&
      typeof out[key] === 'object' &&
      !Array.isArray(out[key])
    ) {
      out[key] = deepMerge(out[key], value);
    } else {
      out[key] = value;
    }
  }
  return out;
}

function gitRoot(startDir) {
  let cur = path.resolve(startDir);
  while (true) {
    if (fs.existsSync(path.join(cur, '.git'))) return cur;
    const parent = path.dirname(cur);
    if (parent === cur) return null;
    cur = parent;
  }
}

export function findPfConfigPath(startDir = process.cwd()) {
  const start = path.resolve(startDir);
  const stop = gitRoot(start) || path.parse(start).root;

  let cur = start;
  while (true) {
    const candidate = path.join(cur, 'pf.config.json5');
    if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) return candidate;

    if (cur === stop) break;
    const parent = path.dirname(cur);
    if (parent === cur) break;
    cur = parent;
  }

  const home = os.homedir();
  const userCandidates = [
    path.join(home, '.config', 'pf', 'pf.config.json5'),
    path.join(home, '.pf', 'pf.config.json5'),
  ];
  for (const candidate of userCandidates) {
    if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) return candidate;
  }

  return null;
}

export function loadPfConfig(startDir = process.cwd()) {
  const configPath = findPfConfigPath(startDir);
  if (!configPath) return { config: { ...DEFAULT_CONFIG }, path: null };

  const raw = fs.readFileSync(configPath, 'utf-8');
  const parsed = JSON5.parse(raw);
  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new Error(`Invalid config root (expected object): ${configPath}`);
  }

  return { config: deepMerge(DEFAULT_CONFIG, parsed), path: configPath };
}

export function getConfigValue(config, dottedKey, defaultValue) {
  let cur = config;
  for (const part of dottedKey.split('.')) {
    if (!cur || typeof cur !== 'object' || Array.isArray(cur) || !(part in cur)) return defaultValue;
    cur = cur[part];
  }
  return cur ?? defaultValue;
}

export function resolveHomePath(value) {
  if (typeof value !== 'string') return value;
  if (value.startsWith('~/')) return path.join(os.homedir(), value.slice(2));
  if (value === '~') return os.homedir();
  return value;
}
